src/dmd/dsymbolsem.d

```diff
@@ -3735,7 +3735,17 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         if (dd.ident == Id.dtor && dd.semanticRun < PASS.semantic)
             ad.dtors.push(dd);
         if (!dd.type)
-            dd.type = new TypeFunction(null, Type.tvoid, false, LINK.d, dd.storage_class);
+        {
+            Parameters* params = null;
+            if (ad.classKind == ClassKind.cpp && !Target.cppDeletingDestructor)
+            {
+                // Windows doesn't use a deleting destructor, it takes an argument instead!
+                Parameter param = new Parameter(STC.undefined_, Type.tuns32, new Identifier("del"), new IntegerExp(Loc.initial, 0, Type.tuns32));
```

Comment:
I don't think it can work so easily (AFAIU you are adding a parameter with a default value so it can still be called without arguments). This helps for explicit calls, but not for the dtor written to the TypeInfo. The latter needs a shim that adds the default value as an argument. Here's an example that calls the dtor this way:
```
import core.stdc.stdio;

extern(C++) 
struct S
{
	int x;
	~this() { printf("~S %x\n", x); }
}

void main()
{
	S[] arr = new S[3];
	arr[1].x = 1;
	arr[2].x = 2;
	arr.length = 1;
	assumeSafeAppend(arr); // destroys arr[1] and arr[2]
	printf("done\n");
}
```
Output is:
```
~S 2
~S 1
done
~S 0
```
Works for Win64, but not for Win32 in dmd 2.080. (Falsely assumes that C++ dtor does not have an argument, though).

Reply:
Oh no, you're killing me bro! >_<

Reply:
Here I was thinking this was a super clever and elegant solution :(

Reply:
Perhaps I could address this in a follow-up? Situation in this PR is immeasurably better than what we had before, and I have people here that want to be using this code already.
Obviously, I'm happy to continue to work on this issue until it's correct.

Reply:
Maybe druntime can handle that case, too. Any user of the dtor member of TypeInfo_Struct/Class needs to know about it then, though.

Reply:
> Perhaps I could address this in a follow-up?

Sounds reasonable.

Reply:
Can you put this snippet in an issue?
I'm not sure if I'll work on this case, or the constructor issue first...

Reply:
> Why is it that druntime isn't calling the destructor with the default arg supplied like everywhere else?

Because it uses a function pointer assuming it does not need an extra argument. See
https://github.com/dlang/druntime/blob/master/src/rt/lifetime.d#L1347
https://github.com/dlang/druntime/blob/master/src/object.d#L2010

> Can you put this snippet in an issue?

https://issues.dlang.org/show_bug.cgi?id=18953

I just realized that calling the C++ dtor is pretty much broken for Win32 anyway because it doesn't use the __thiscall ABI.

Reply:
We'll need that hack you suggest to generate a shim when assigning to a delegate I guess.